### ISSUE TITLE
Add TypeScript definitions

### ISSUE THREAD

**Issue by mekwall** [sonaye/mobx-apollo#7]
Since mobx is written in TypeScript and comes bundle with it's definition file it would be nice to have it for mobx-apollo as well.

**Comment by sonaye** [comment]
Unfortunately I am not very familiar with TS, I chose to go with Flow when I started investigating a better typed JS. What kind of advantages can we gain by rewriting such a small package in TS may I ask?

**Comment by mekwall** [comment]
You don't have to rewrite it. Just add the definitions file. To get some inspiration you can take a look at how that's done here: https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types

**Comment by sonaye** [comment]
I truly cannot help with that, I have zero knowledge about TS and don't have the time to invest learning it. You can submit a PR if you wish and think it's vital to have, can you at least tell me about the value it adds to the package?

**Comment by pselden** [comment]
I'm using this in Typescript. I'll see If I can come up with some typedefs. The main problem is that you cannot use the decorator version because typescript does not support the descriptor parameter in property decorators.

**Comment by sonaye** [comment]
@pselden Are you saying that the decorated version should be removed from the package -in order for typescript defs to be added- or that only definition for the non decorated version can be added?

**Comment by pselden** [comment]
I guess my statement was tangential to this discussion, but when using TypeScript you cannot have an initializer with a property decorator, so the decorator just cannot be used at all.

```ts
class Store() {
   @query
   posts: PostsQueryObservable = { client, query: POSTS_QUERY}; // the client/query object doesn't even get passed to the decorator, so it just has a runtime error that descriptor is undefined
}
```

You can still use the non-decorator syntax. To use the decorator in typescript you'd have to create a decorator factory instead.

```js
function queryDecorator(descriptor) {
   return function(target, propName) {
      query(target, propName, queryDecorator);
   };
}
```

All that said, it wouldn't really stop us from making type defs for this, just makes it harder to use in TypeScript.

**Comment by sonaye** [comment]
@pselden Interesting. I still cannot figure out the advantages of adding type defs to such a small package, I can understand it for something more complex. Can you shed some light on this?

**Comment by pselden** [comment]
It's sort of more useful now that the returned value is not _just_ the props types.

Here's my WIP for the types I'm using internally.

```ts
declare module 'mobx-apollo' {
  import {
    ApolloClient,
    ObservableQuery,
    WatchQueryOptions
  } from 'apollo-client';
  
  export declare function query(
    target: Object,
    propName: string,
    descriptor: {
      client: ApolloClient;
    } & WatchQueryOptions
  );

  export interface MobxApolloQuery<T> {
    loading: boolean;
    error?: ApolloError;
    data: T;
    ref: ObservableQuery<T>;
  }
}
```

So now I import the MobxApolloQuery type and can use it like:

```ts
class Store {
   private organizationsQuery: MobxApolloQuery<Organizations>;

   constructor() {
      query(this, 'organizationsQuery', {
        client: client; // makes sure i'm passing the right kind of client,
      });
   }

   @computed
   get organizations() {
      return this.organizationsQuery.data.organizations.edges; // i get type support here (intellisense from my editor, type checking during compile so I make sure I didn't make a typo).
   }
}
```

**Comment by VojtechStep** [comment]
@sonaye To address your question that everybody here apparently refuses to answer 😄 acknowledge:
TypeScript definitions for your package would provide better programming experience for programmers using this package in TypeScript or even in JavaScript if they use an IDE that supports intellisense in JS based on TypeScript definition files (the only one I'm aware of at the moment is Visual Studio Code).

If the definitions file is provided, typescript won't transpile code that violates the defined structure, for example when you make a typo or assume a return type of a function without reading the documentation.

Basically instead of having to read the docs or explore the code, you can instantly know what properties you can pass to a function or expect to receive in return. A typescript definition file represents a contract between the library developer and the app developer.

**Comment by pselden** [comment]
Circling back to this -- we have two options:

1) Add types directly to this repository -- @sonaye and others would have the responsibility of maintaining it here. I've seen projects where the primary maintainer does not use typescript yet includes community provided typescript definitions (https://github.com/blainekasten/enzyme-matchers), usually the maintainer doesn't bother updating the types and just has people fix them afterwards. Types are automatically included when using this project.

2) The community adds their own types to https://github.com/DefinitelyTyped/DefinitelyTyped -- it essentially has no relation to this project. It's up to the community (likely whoever added it in the first place, to maintain or accept PRs when the original project updates. TypeScript uses can then use it by installing @types/mobx-apollo.

So the question is I suppose whether @sonaye is willing to accept slightly more burden for maintaining this project to accept additional PRs (and actually most likely people will just create a GH issue and not do a PR) to update the types if/when the main code changes.

**Comment by sonaye** [comment]
I am ok with both, I just cannot provide the types myself because I don't use Typescript, I personally use Flow. That being said, I would prefer 2 since there is already an ecosystem for it out there.

**Comment by pselden** [comment]
Here's what I threw together for the new API:

```ts
declare module 'mobx-apollo' {
  import {
    ApolloClient,
    ObservableQuery,
    WatchQueryOptions,
    ApolloError
  } from 'apollo-client';

  interface MobxApolloQueryOptions<T> extends WatchQueryOptions {
    client: ApolloClient;
    onFetch?: (result: T) => void;
    onError?: (error: ApolloError) => void;
  }

  interface MobxApolloQuery<T> {
    loading: boolean;
    data: T;
    error?: ApolloError;
    ref: ObservableQuery<T>;
  }

  function graphql<T>(options: MobxApolloQueryOptions<T>): MobxApolloQuery<T> {
  }

  export default graphql;
}
```

I'll look into adding something similar to DefinitelyTyped.

**Comment by sonaye** [comment]
@pselden Great work!

**Comment by mekwall** [comment]
Thanks! Sorry I couldn't contribute more myself, but happy that others could 👍
